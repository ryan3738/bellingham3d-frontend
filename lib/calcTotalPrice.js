export default function calcTotalPrice(cart) {
  return cart.reduce((tally, cartItem) => {
    // products can be deleted, but they could still be in your cart
    if (!cartItem.product) return tally;
    return tally + cartItem.quantity * cartItem.product.price;
  }, 0);
}
