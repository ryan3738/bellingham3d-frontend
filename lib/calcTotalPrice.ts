export default function calcTotalPrice(cart: any): number {
  return cart.reduce((tally, cartItem) => {
    // products can be deleted, but they could still be in your cart
    if (!cartItem.product) return tally;
    return tally + cartItem.quantity * cartItem.product.price;
  }, 0);
}
