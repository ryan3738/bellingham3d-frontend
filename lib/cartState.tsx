import { createContext, useContext, useState } from 'react';

interface CartContextInterface {
  cartOpen: boolean;
  setCartOpen: (boolean) => void;
  toggleCart: () => void;
  closeCart: () => void;
  openCart: () => void;
}
const CartContext = createContext<CartContextInterface | null>(null);
const CartProvider = CartContext.Provider;

type AppProps = {
  children?: React.ReactNode;
};

function CartStateProvider({ children }: AppProps): JSX.Element {
  // This is our own custom provider! We will store data (state) and functionality (updaters) in here and anyone can access it via the consumer!

  // Close by default
  const [cartOpen, setCartOpen] = useState(false);

  function toggleCart(): void {
    setCartOpen(!cartOpen);
  }

  function closeCart(): void {
    setCartOpen(false);
  }

  function openCart(): void {
    setCartOpen(true);
  }

  return (
    <CartProvider
      value={{
        cartOpen,
        setCartOpen,
        toggleCart,
        closeCart,
        openCart,
      }}
    >
      {children}
    </CartProvider>
  );
}

// make a custom hook for accessing the cart
function useCart(): CartContextInterface {
  // We use a consumer here to access the local state
  const all = useContext(CartContext);
  return all;
}

export { CartStateProvider, useCart };
