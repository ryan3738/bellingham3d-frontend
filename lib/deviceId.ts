import { v4 as uuidv4 } from 'uuid';

const getDeviceId = (): string => {
  const myuuid = uuidv4();
  console.log(`Your UUID is: ${myuuid}`);

  return uuidv4();
};

export { getDeviceId };
