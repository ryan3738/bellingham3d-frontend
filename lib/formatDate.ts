interface DateTimeFormatOptions {
  localeMatcher?: 'best fit' | 'lookup';
  weekday?: 'long' | 'short' | 'narrow';
  era?: 'long' | 'short' | 'narrow';
  year?: 'numeric' | '2-digit';
  month?: 'numeric' | '2-digit' | 'long' | 'short' | 'narrow';
  day?: 'numeric' | '2-digit';
  hour?: 'numeric' | '2-digit';
  minute?: 'numeric' | '2-digit';
  second?: 'numeric' | '2-digit';
  timeZoneName?: 'long' | 'short';
  formatMatcher?: 'best fit' | 'basic';
  hour12?: boolean;
  timeZone?: string;
}

const formatDate = (dateString: string): string => {
  const options: DateTimeFormatOptions = {
    year: 'numeric',
    month: 'long',
    day: 'numeric',
  };
  return new Date(dateString).toLocaleDateString(undefined, options);
};

export { formatDate };
