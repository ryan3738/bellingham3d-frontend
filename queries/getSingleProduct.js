import gql from 'graphql-tag';

export const SINGLE_PRODUCT_QUERY = gql`
  query SINGLE_PRODUCT_QUERY($id: ID!) {
    product(where: { id: $id }) {
      id
      name
      price
      description
      variants {
        id
        name
        variantType {
          id
          name
        }
      }
      image {
        altText
        image {
          publicUrlTransformed(transformation: { width: "600" })
        }
      }
    }
  }
`;
