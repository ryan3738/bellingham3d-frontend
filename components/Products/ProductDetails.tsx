import { useQuery } from '@apollo/client';
import Head from 'next/head';
import { useCallback, useState } from 'react';
import styled from 'styled-components';
import formatMoney from '../../lib/formatMoney';
import AddToCart from '../Cart/AddToCart';
import DisplayApolloError from '../DisplayApolloError';
import ImageSlider from '../ImageSlider';
import ProductVariants from './ProductVariants';
import { useUser } from '../User/User';
import { SINGLE_PRODUCT_QUERY } from '../../queries/getSingleProduct';
import SeeAllProducts from './SeeAllProducts';
import { siteData } from '../../public/site-data';
import {
  Option,
  ProductType,
  SelectVariantType,
  Variant,
} from '../../types/types';
import AuthTabs from '../User/AuthTabs';

const ProductStyles = styled.div`
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(240px, 1fr));
  align-content: center;
  justify-content: space-evenly;
  justify-items: center;
  align-items: start;
  width: 100%;
  height: auto;
  gap: 2rem;
  img {
    width: 100%;
    object-fit: contain;
  }
`;

type AppProps = {
  id: ProductType['id'];
};

interface SelectedVariant extends Option {
  variant: Variant;
}

export default function ProductDetails({ id }: AppProps): JSX.Element {
  const [variantsState, setVariantsState] = useState<SelectedVariant[]>([]);
  const user = useUser();
  console.log('ID', id);

  const selectVariant: SelectVariantType = useCallback(
    ({ option, variant }) => {
      // Take in options and variants
      setVariantsState((previousState) => {
        // If the option doesn't exist add it to state
        if (!previousState.find((item) => item.id === option.id)) {
          const newOption = {
            id: option.id,
            name: option.name,
            variant: {
              id: variant.id,
              name: variant.name,
            },
          };
          return [...previousState, newOption];
        }
        previousState.forEach((item) => {
          // If option exists, update the variant state
          if (item.name === option.name) {
            item.variant.id = variant.id;
            item.variant.name = variant.name;
          }
        });
        return [...previousState];
      });
    },
    []
  );

  const { data, loading, error } = useQuery(SINGLE_PRODUCT_QUERY, {
    variables: {
      id,
    },
  });
  if (loading) return <p>Loading...</p>;
  if (error) return <DisplayApolloError error={error} />;
  if (!data) return <SeeAllProducts />;

  const { product } = data || null;
  const getVariantIds = (options: SelectedVariant[]): Variant['id'][] => {
    const variantIds = options.map((option) => option.variant.id);
    // console.log('variantIds', variantIds);
    return variantIds;
  };

  return (
    <>
      <ProductStyles>
        <Head>
          <title>
            {product.name} | {siteData.businessName}
          </title>
        </Head>
        <ImageSlider slides={product.images} alt={product.name} />
        <div className="details">
          <h2>{product.name}</h2>
          <h3>{formatMoney(product.price)}</h3>
          <ProductVariants
            variants={product.variants}
            selectVariant={selectVariant}
          />

          <AddToCart
            id={product.id}
            variantIds={getVariantIds(variantsState)}
            disabled={!user}
          />

          <p>{product.description}</p>
          {!user && (
            <>
              <div>
                <h3>You must be signed in to add items to your cart</h3>
              </div>
              <AuthTabs />
            </>
          )}
          <SeeAllProducts />
        </div>

        <style jsx>{`
          select {
            width: 100%;
            padding: 0.5rem;
            font-size: 1rem;
            border: 1px solid black;
            margin: 5px 0 20px 0;
          }
          select:focus {
            outline: 0;
            border-color: var(--navyBlue);
          }
        `}</style>
      </ProductStyles>
    </>
  );
}
