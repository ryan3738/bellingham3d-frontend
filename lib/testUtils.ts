import casual from 'casual';
import { PAGINATION_QUERY } from '../components/Products/Pagination';
import {
  CartItemType,
  OrderItemType,
  OrderType,
  ProductType,
  UserType,
} from '../types/types';

// seed it so we get consistent results
casual.seed(777);

const fakeProduct = (): ProductType => ({
  __typename: 'Product',
  id: 'abc123',
  price: 5000,
  // status: 'AVAILABLE',
  variants: [],
  images: [
    {
      __typename: 'ProductImage',
      // id: 'abc123',
      altText: 'dogs are best',
      image: {
        __typename: 'CloudinaryImage_File',
        publicUrlTransformed: '/dog.jpg',
      },
    },
  ],
  name: 'dogs are best',
  description: 'dogs',
});

const fakeUser = (overrides?: UserType): UserType => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  role: ['ADMIN'],
  // orders: [],
  cart: [],
  ...overrides,
});

const fakeOrderItem = (): OrderItemType => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: {
    __typename: 'ProductImage',
    id: 'abc123',
    altText: 'dogs are best',
    image: {
      id: 'abc123',
      __typename: 'CloudinaryImage_File',
      publicUrlTransformed: `${casual.word}.jpg`,
    },
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = (): OrderType => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2022-12-11T20:16:13.797Z',
  shippingAddress: {
    __typename: 'CustomerAddress',
    id: casual.uuid,
    firstName: casual.first_name,
    lastName: casual.last_name,
    company: casual.company_name,
    address1: casual.address1,
    address2: casual.address2,
    city: casual.city,
    region: casual.state_abbr,
    country: casual.country,
    zip: casual.zip({}),
    phone: casual.phone,
  },
  // user: fakeUser(),
});

const fakeCartItem = (overrides?: CartItemType): CartItemType => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  product: fakeProduct(),
  // user: fakeUser(),
  ...overrides,
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

function makePaginationMocksFor(length) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: {
        data: {
          _allProductsMeta: {
            count: length,
          },
          itemsConnection: {
            __typename: 'aggregate',
            aggregate: {
              count: length,
              __typename: 'count',
            },
          },
        },
      },
    },
  ];
}

export {
  makePaginationMocksFor,
  LocalStorageMock,
  fakeProduct,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
};
