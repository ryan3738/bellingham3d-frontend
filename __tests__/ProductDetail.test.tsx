import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import { fakeProduct } from '../lib/testUtils';
import { SINGLE_PRODUCT_QUERY } from '../queries/getSingleProduct';
import ProductDetails from '../components/Products/ProductDetails';

const product = fakeProduct();

const mocks = [
  {
    // When somoen requests this query and variable combo
    request: {
      query: SINGLE_PRODUCT_QUERY,
      variables: {
        id: '123',
      },
    },
    // Return this data
    result: {
      data: {
        product,
      },
    },
  },
];

describe('<ProductDetails />', () => {
  it('renders with proper data', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <ProductDetails id="123" />
      </MockedProvider>
    );
    // Wait for the test ID to show up
    await screen.findByTestId('productDetails');
    debug();
  });
});
