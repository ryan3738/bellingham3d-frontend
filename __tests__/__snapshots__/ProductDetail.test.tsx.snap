// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDetails /> Errors out when an item is not found 1`] = `
<div>
  <div
    class="DisplayApolloError__ErrorStyles-sc-yc1ztr-0 iSnxOG"
  >
    <p
      data-testid="graphql-error"
    >
      <strong>
        Error!!
      </strong>
      Item not found
    </p>
  </div>
</div>
`;

exports[`<ProductDetails /> renders with proper data 1`] = `
<div>
  <div
    class="ProductDetails__ProductStyles-sc-xqwt7f-0 htxGUr"
    data-testid="productDetails"
  >
    <div
      class="jsx-1005620270 slider"
    >
      <div
        class="jsx-1005620270 slide active"
      >
        <span
          style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
        >
          <span
            style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27600%27%20height=%27600%27/%3e"
              style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
            />
          </span>
          <img
            alt="dogs are best"
            data-nimg="intrinsic"
            decoding="async"
            src="/_next/image?url=%2Fdog.jpg&w=1200&q=75"
            srcset="/_next/image?url=%2Fdog.jpg&w=640&q=75 1x, /_next/image?url=%2Fdog.jpg&w=1200&q=75 2x"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
      </div>
    </div>
    <div
      class="jsx-550922915 details"
    >
      <h2
        class="jsx-550922915"
      >
        dogs are best
      </h2>
      <p
        class="jsx-550922915"
      >
        dogs
      </p>
      <h3
        class="jsx-550922915"
      >
        $50
      </h3>
      <div
        class="details"
      />
      <button
        class="StateStyles__ButtonStyles-sc-1rtfr0y-0 fTnNMI"
        disabled=""
        type="button"
      >
        Add
         To Cart
      </button>
      <div
        class="jsx-550922915"
      >
        <strong
          class="jsx-550922915"
        >
          You must be signed in to add items to your cart
        </strong>
      </div>
      <div
        class="AuthTabs__TabContainerStyles-sc-ovdjsg-0 brbQT"
      >
        <div
          class="AuthTabs__TabButtonsStyles-sc-ovdjsg-1 BhBkv"
        >
          <button
            class="StateStyles__ButtonStyles-sc-1rtfr0y-0 StateStyles__TabStyles-sc-1rtfr0y-1 fTnNMI gNeQwd"
            name="default"
            type="button"
            value="signup"
          >
            Sign Up
          </button>
          <button
            class="StateStyles__ButtonStyles-sc-1rtfr0y-0 StateStyles__TabStyles-sc-1rtfr0y-1 fTnNMI gNeQwd"
            name="select"
            type="button"
            value="signin"
          >
            Sign In
          </button>
        </div>
        <div />
      </div>
    </div>
  </div>
</div>
`;
